--- api_client.mustache	2024-08-13 14:29:04.056364916 -0500
+++ api_client_new.mustache	2024-08-13 14:29:36.224410735 -0500
@@ -159,6 +159,7 @@
 {{#native_serialization}}
   /// Returns a native instance of an OpenAPI class matching the [specified type][targetType].
   static dynamic fromJson(dynamic value, String targetType, {bool growable = false,}) {
+    upgradeDto(value, targetType);
     try {
       switch (targetType) {
         case 'String':
