--- api_client.mustache	2024-08-13 14:12:55
+++ api_client1.mustache	2024-08-13 14:12:01
@@ -159,7 +159,6 @@
 {{#native_serialization}}
   /// Returns a native instance of an OpenAPI class matching the [specified type][targetType].
   static dynamic fromJson(dynamic value, String targetType, {bool growable = false,}) {
-    upgradeDto(value, targetType);
     try {
       switch (targetType) {
         case 'String':
