--- native_class.mustache	2023-06-22 12:00:16.480685565 -0500
+++ native_class1.mustache	2023-06-22 12:01:41.772534371 -0500
@@ -91,14 +91,14 @@
     {{/isDateTime}}
     {{#isNullable}}
     } else {
-      json[r'{{{baseName}}}'] = null;
+      // json[r'{{{baseName}}}'] = null;
     }
     {{/isNullable}}
     {{^isNullable}}
       {{^required}}
         {{^defaultValue}}
     } else {
-      json[r'{{{baseName}}}'] = null;
+      // json[r'{{{baseName}}}'] = null;
     }
         {{/defaultValue}}
       {{/required}}
@@ -120,7 +120,7 @@
       assert(() {
         requiredKeys.forEach((key) {
           assert(json.containsKey(key), 'Required key "{{{classname}}}[$key]" is missing from JSON.');
-          assert(json[key] != null, 'Required key "{{{classname}}}[$key]" has a null value in JSON.');
+          // assert(json[key] != null, 'Required key "{{{classname}}}[$key]" has a null value in JSON.');
         });
         return true;
       }());
